vulnerability/replicator_test.go

```diff
@@ -0,0 +1,113 @@
+	"sync"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/elastic/cloudbeat/resources/providers/awslib/ec2"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/stretchr/testify/assert"
+)
+
+type mockProviderReplicator struct{}
+
+func (p *mockProviderReplicator) CreateSnapshots(ctx context.Context, data ec2.Ec2Instance) ([]ec2.EBSSnapshot, error) {
+	// Mock implementation of CreateSnapshots function
+	// Return a slice of ec2.EBSSnapshot for testing purposes
+	return []ec2.EBSSnapshot{
+		{Instance: data, SnapshotId: "snapshot-1"},
+		{Instance: data, SnapshotId: "snapshot-2"},
+		{Instance: data, SnapshotId: "snapshot-3"},
+	}, nil
+}
+
+func TestVulnerabilityReplicator_SnapshotInstance(t *testing.T) {
+	log := logp.NewLogger("TestVulnerabilityReplicator_SnapshotInstance")
+
+	// Create a mock provider
+	provider := &mockProviderReplicator{}
+
+	// Create an instance of VulnerabilityReplicator
+	replicator := NewVulnerabilityReplicator(log, provider)
+
+	// Create a context for the test
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create a channel for sending Ec2Instance objects
+	insCh := make(chan ec2.Ec2Instance)
+
+	wg := sync.WaitGroup{}
+	// Start the SnapshotInstance goroutine
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		replicator.SnapshotInstance(ctx, insCh)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// Collect the received snapshots from the channel
+		var receivedSnapshots []ec2.EBSSnapshot
+		for s := range replicator.GetChan() {
+			receivedSnapshots = append(receivedSnapshots, s)
+		}
+
+		// Assert that the received snapshots match the expected snapshots
+		expectedSnapshots := []ec2.EBSSnapshot{
+			{Instance: ec2.Ec2Instance{
+				Instance: types.Instance{InstanceId: aws.String("instance-1")},
+			}, SnapshotId: "snapshot-1"},
+			{Instance: ec2.Ec2Instance{
+				Instance: types.Instance{InstanceId: aws.String("instance-1")},
+			}, SnapshotId: "snapshot-2"},
+			{Instance: ec2.Ec2Instance{
+				Instance: types.Instance{InstanceId: aws.String("instance-1")},
+			}, SnapshotId: "snapshot-3"},
+			{Instance: ec2.Ec2Instance{
+				Instance: types.Instance{InstanceId: aws.String("instance-2")},
+			}, SnapshotId: "snapshot-1"},
+			{Instance: ec2.Ec2Instance{
+				Instance: types.Instance{InstanceId: aws.String("instance-2")},
+			}, SnapshotId: "snapshot-2"},
+			{Instance: ec2.Ec2Instance{
+				Instance: types.Instance{InstanceId: aws.String("instance-2")},
+			}, SnapshotId: "snapshot-3"},
+		}
+
+		assert.ElementsMatch(t, expectedSnapshots, receivedSnapshots)
+	}()
+
+	// Send Ec2Instance data to the channel
+	insCh <- ec2.Ec2Instance{
+		Instance: types.Instance{InstanceId: aws.String("instance-1")},
+	}
+	insCh <- ec2.Ec2Instance{
+		Instance: types.Instance{InstanceId: aws.String("instance-2")},
+	}
+
+	// Close the channel to indicate that no more data will be sent
+	close(insCh)
+	wg.Wait()
```

Comment:
A more clean solution for tests is `goleak.VerifyNone`, you can see other places we use it.